Include dataset hash in job search

mvdbeek opened the pull request:
This means we find equivalent jobs if an input hda either points at the same dataset id (existing behavior), or if the dataset ~~source_uri, transform and~~ hashes match. All further restrictions still apply (same metadata etc).

Builds on #19108 and https://github.com/galaxyproject/galaxy/pull/19110 .

Also:
- Drop `test_run_deferred_dataset` and `test_run_deferred_dataset_with_metadata_options_filter`
  tests from `lib/galaxy_test/api/test_tools.py` which are duplicates
  of `test_deferred_with_cached_input` and `test_deferred_with_metadata_options_filter` in
  `lib/galaxy_test/api/test_tool_execute.py` .
- Improve type annotations

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

mvdbeek committed:
Include dataset hash in job search

nsoranzo commented:
Rebase?

bgruening commented:
Running a Galaxy Training Academy without large computational resources - yeah, yeah!

jmchilton commented:
We have no clue about extra files in this context right - how can we make this conclusion without that? It isn't a -1 but I am uncomfortable with the direction this is all heading. I feel like we should be working on tightening up job search and not loosening it more and more before we've installed the guard rails it needs IMO.

nsoranzo commented on lib/galaxy_test/api/test_tools.py:
```diff
@@ -2436,22 +2436,45 @@ def test_group_tag_selection_multiple(self, history_id):
-    @skip_without_tool("cat1")
-    def test_run_deferred_dataset(self, history_id):
+    def _run_deferred(self, history_id: str, use_cached_job=False, expect_cached_job=False, include_correct_hash=True):
+        hashes: Optional[List[Dict[str, str]]] = None
+        if include_correct_hash:
+            hashes = [{"hash_function": "SHA-1", "hash_value": "2d7dcdb10964872752bd6d081725792b3f729ac9"}]
```
You can skip the explicit type annotation with:
```suggestion
        hashes = (
            [{"hash_function": "SHA-1", "hash_value": "2d7dcdb10964872752bd6d081725792b3f729ac9"}]
            if include_correct_hash
            else None
        )
```

mvdbeek commented:
>  I would implement a hash of the dataset and not of the primary file of the dataset. 95% of the time they could be the same but we should verify that before using it in this context.

If we match the transform would it still be possible to get different extra files for uploaded content (considering we already match on the datatype) ? I added https://github.com/galaxyproject/galaxy/pull/19112/commits/4abc6e60207ea34792c2839a32292b82e7ac8bbc because we don't record the transform for local files, but we could of course do that.

There's of course also value in calculating the hash for datasets as they are written to the object store, but if we can get reliable equivalence using the upload hash + transform we could make nice progress in figuring out other edge cases for IWC workflows ?

jmchilton commented:
I had the realization reading your response that you only care about upload or think these fields can all only be set during uploads. I believe any tool can produce hashes and source hashes and I believe we have an API for hashing files after the fact anyway. Could we restrict all this logic to just fetch and upload1 - then it feels pretty close to being correct? This also probably explains my unease with #19110 - I think we don't validate the hashes outside the fetch tool but we can create the hashes in other tools I think - it feels like what we need is a hash validated field if we want to act on that data in this fashion but maybe it would be sufficient to be more proactive in validating the hash field whenever it is set.

nsoranzo commented:
You may want to rebase this now that #19181 has been merged (thanks @jmchilton !), since it contained some type annotation fixes copied from here.

> If we match the transform would it still be possible to get different extra files for uploaded content (considering we already match on the datatype) ?

Not sure it's the same problem, but in #19181 I found out that a type of transform (grooming) is not reproducible (because the BAM header contains paths as part of the samtools command used to sort the data), so you'd get different hashes if you materialize twice from the same source.

So I think I agree it's safer to match on the actual dataset hashes instead of the dataset source hashes.

mvdbeek commented:
Uhm, I'm confused, is that not what I'm doing ?

nsoranzo commented:
@mvdbeek Still draft?

mvdbeek commented:
I think so ? I need to think about all the ways it could break, and extra files is a valid concern, I think ?

mvdbeek commented:
That sounds good to me. I'm hacking away on something else right now, if you want to add to the PR that would be really cool, otherwise I'll come back to it later.

mvdbeek commented on lib/galaxy/managers/jobs.py:
```diff
@@ -137,6 +137,31 @@ def safe_aliased(model_class, name=None):
+    a_hash = aliased(model.DatasetHash)
+    b_hash = aliased(model.DatasetHash)
+    stmt = (
+        stmt.outerjoin(a_hash, a.dataset_id == a_hash.dataset_id)
```
The outerjoin is likely going to be really bad for performance

nsoranzo committed:
Enable hash-based job matching for composite datasets

by computing and storing hashes for all files (primary file and extra files).

Enhanced ``has_same_hash()`` to match datasets only when all hashes match,
preventing partial matches.
